Pinta.Core/Effects/ColorGradient.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Pinta.Core;
+
+public sealed class ColorGradient : ColorMapping
+{
+	public ColorBgra StartColor { get; }
+	public ColorBgra EndColor { get; }
+	public double MinimumPosition { get; }
+	public double MaximumPosition { get; }
+
+	private readonly IReadOnlyList<GradientStop> sorted_stops; // TODO: Before making it public think of nice structures
+
+	internal ColorGradient (
+		ColorBgra startColor,
+		ColorBgra endColor,
+		double minPosition,
+		double maxPosition,
+		IEnumerable<KeyValuePair<double, ColorBgra>> gradientStops)
+	{
+		if (minPosition >= maxPosition) throw new ArgumentException ($"{nameof (minPosition)} has to be lower than {nameof (maxPosition)}");
+		var sortedStops = gradientStops.Select (kvp => new GradientStop (kvp.Key, kvp.Value)).OrderBy (stop => stop.Position).ToArray ();
+		if (sortedStops.Length > 0 && sortedStops[0].Position <= minPosition) throw new ArgumentException ($"Lowest key in {nameof (gradientStops)} has to be greater than {nameof (minPosition)}");
+		if (sortedStops.Length > 0 && sortedStops[^1].Position >= maxPosition) throw new ArgumentException ($"Greatest key in {nameof (gradientStops)} has to be lower than {nameof (maxPosition)}");
+		var distinctPositions = sortedStops.GroupBy (s => s.Position).Count ();
+		if (distinctPositions != sortedStops.Length) throw new ArgumentException ("Cannot have more than one stop in the same position");
+
+		StartColor = startColor;
+		EndColor = endColor;
+		MinimumPosition = minPosition;
+		MaximumPosition = maxPosition;
+		sorted_stops = sortedStops;
+	}
+
+	public sealed override ColorBgra GetColor (double position)
+	{
+		if (position == MinimumPosition) return StartColor;
```

Comment:
If the list of `sorted_stops` includes the start / end colours, that might unify some of these branches and simplify the logic a bit?